src/libraries/System.Net.Mail/src/System/Net/Mime/QEncodedStream.cs

```diff
@@ -221,38 +217,149 @@ public int EncodeBytes(byte[] buffer, int offset, int count)
-                    WriteState.Append((byte)HexConverter.ToCharUpper(buffer[cur] >> 4));
-                    //clear the first four bytes to get the last four and look up the hex digit
-                    WriteState.Append((byte)HexConverter.ToCharUpper(buffer[cur]));
+                    ApppendEncodedByte(buffer[cur]);
                 }
             }
             WriteState.AppendFooter();
             return cur - offset;
         }
 
+        public int EncodeString(string value, Encoding encoding)
+        {
+            Debug.Assert(value != null, "value was null");
+            Debug.Assert(_writeState != null, "writestate was null");
+            Debug.Assert(_writeState.Buffer != null, "writestate.buffer was null");
+
+            if (encoding == Encoding.Latin1) // we don't need to check for codepoints
+            {
+                byte[] buffer = encoding.GetBytes(value);
+                return EncodeBytes(buffer, 0, buffer.Length);
+            }
+
+            // Add Encoding header, if any. e.g. =?encoding?b?
+            WriteState.AppendHeader();
+
+            int totalBytesCount = 0;
+            byte[] bytes = new byte[encoding.GetMaxByteCount(2)];
+            for (int i = 0; i < value.Length; ++i)
+            {
+                int codepointSize = GetCodepointSize(value, i);
+                Debug.Assert(codepointSize == 1 || codepointSize == 2, "codepointSize was not 1 or 2");
+
+                int bytesCount = encoding.GetBytes(value, i, codepointSize, bytes, 0);
+                if (codepointSize == 2)
+                {
+                    ++i; // Transformed two chars, so shift the index to account for that
+                }
+                AppendEncodedCodepoint(bytes, bytesCount);
+                totalBytesCount += bytesCount;
+            }
+
+            // Write out the last footer, if any.  e.g. ?=
+            WriteState.AppendFooter();
+
+            return totalBytesCount;
+        }
+
+        private static int GetCodepointSize(string value, int i)
+        {
+            // specific encoding for CRLF
+            if (value[i] == '\r' && i + 1 < value.Length && value[i + 1] == '\n')
+            {
+                return 2;
+            }
+
+            if (char.IsSurrogate(value[i]) && i + 1 < value.Length && char.IsSurrogatePair(value[i], value[i + 1]))
+            {
+                return 2;
+            }
+
+            return 1;
+        }
+
+        private int AppendEncodedCodepoint(byte[] bytes, int count)
+        {
+            if (LineBreakNeeded(bytes, count))
+            {
+                WriteState.AppendCRLF(true);
+            }
+
+            // specific encoding for CRLF
+            if (IsCRLF(bytes, count))
+            {
+                AppendEncodedCRLF();
+                return 2;
+            }
+
+            for (int i = 0; i < count; ++i)
+            {
+                ApppendEncodedByte(bytes[i]);
+            }
+            return count;
+        }
```

Comment:
All this logic looks very similar to the [one in Base64Stream](https://github.com/dotnet/runtime/blob/cf56c4052bf52d7b7e2e85131d23874c88759ec2/src/libraries/System.Net.Mail/src/System/Net/Base64Stream.cs#L222) except for CRLF handling. Wouldn't it be better to extract it to helper methods and control CRLF handling with arguments? It would greatly reduce code duplication.

Reply:
Fixed